src/hotspot/share/gc/serial/serialCompressor.cpp

```diff
@@ -0,0 +1,825 @@
+  uint _num_spaces;
+
+  // The index of the current space to compact into.
+  uint _index;
+
+  // For a given heap address, compute the index of the
+  // corresponding block in the table.
+  inline size_t addr_to_block_idx(HeapWord* addr) const {
+    assert(addr >= _covered.start() && addr <= _covered.end(), "address must be in heap");
+    return pointer_delta(addr, _covered.start()) >> log_words_per_block();
+  }
+
+  // Given a heap word (usually the start of an object), compute the forwarding address.
+  inline HeapWord* forwardee(HeapWord* addr) const {
+    assert(_mark_bitmap.is_marked(addr), "must be marked");
+    HeapWord* block_base = align_down(addr, bytes_per_block());
+    size_t block = addr_to_block_idx(addr);
+    assert(_bot[block] != nullptr, "must have initialized BOT entry");
+    HeapWord* fwd = _bot[block] + _mark_bitmap.count_marked_words_in_block(block_base, addr);
+    assert(SerialHeap::heap()->is_in_reserved(fwd), "forward addresses must be in heap: addr: " PTR_FORMAT ", fwd: " PTR_FORMAT ", block: " SIZE_FORMAT ", bot[block]: " PTR_FORMAT ", block_base: " PTR_FORMAT, p2i(addr), p2i(fwd), block, p2i(_bot[block]), p2i(block_base));
+    return fwd;
+  }
+
+#ifdef ASSERT
+  // Clear table (only required for assertion in forwardee()).
+  void clear(HeapWord* from, HeapWord* to) {
+    size_t from_block = addr_to_block_idx(from);
+    size_t to_block = addr_to_block_idx(align_up(to, bytes_per_block()));
+    Copy::fill_to_words(reinterpret_cast<HeapWord*>(&_bot[from_block]), to_block - from_block);
+  }
+#endif // ASSERT
+
+  // Get current compaction-top for space at index.
+  HeapWord* get_compaction_top(uint index) const {
+    return _spaces[index]._compaction_top;
+  }
+
+  // Get space at index.
+  ContiguousSpace* get_space(uint index) const {
+    return _spaces[index]._space;
+  }
+
+  // Get first dead word of space at index.
+  HeapWord* get_first_dead(uint index) const {
+    return _spaces[index]._first_dead;
+  }
+
+  // Records first dead word of space at index.
+  void record_first_dead(uint index, HeapWord* first_dead) {
+    assert(_spaces[index]._first_dead == nullptr, "should write only once");
+    _spaces[index]._first_dead = first_dead;
+  }
+
+  // Find the start of the first object at or after addr (but not after limit).
+  HeapWord* first_object_in_block(HeapWord* addr, HeapWord* start, HeapWord* limit) {
+    if (addr >= limit) return limit;
+    if (!_mark_bitmap.is_marked(addr)) {
+      // Easy: find next marked address.
+      return _mark_bitmap.get_next_marked_addr(addr, limit);
+    } else {
+      // Find beginning of live chunk.
+      HeapWord* current = _mark_bitmap.get_last_unmarked_addr(start, addr) + 1;
+      // Forward-search to first object >= addr.
+      while (current < addr) {
+        size_t obj_size = cast_to_oop(current)->size();
+        current += obj_size;
+      }
+      assert(current >= addr, "found object start must be >= addr");
+      return addr;
```

Comment:
Should return `MIN2(current, limit)`, no?

Reply:
Duh. Yes, of course! Thanks for pointing it out!